### PR TITLE
Alphabet Search Enhancement - Defines parameters within options

### PR DESCRIPTION
**Pull request by PrzemyslawKlys**
This PR changes how to define options for AlphabetSearch.

Example: https://jsfiddle.net/MadBoy/Lu51s0vc/3/

Instead

```js
$.fn.dataTable.AlphabetSearch.column = 5;
$.fn.dataTable.AlphabetSearch.caseSensitive = true;
$.fn.dataTable.AlphabetSearch.addNumbers = true;
```

New options - which seems a bit nicer and easier to set and follow

```
alphabet: {
    column: 3,
    caseSensitive: true,
    numbers: true,
},
```

```js
<script>
    $(document).ready(function () {
        $.fn.dataTable.moment("L");
        //$.fn.dataTable.AlphabetSearch.column = 5;
        //$.fn.dataTable.AlphabetSearch.caseSensitive = true;
        //$.fn.dataTable.AlphabetSearch.addNumbers = true;
        var table = $("#DT-rzdmiDVT").DataTable({
            dom: "ABfrtip",
            searchFade: !1,
            scrollCollapse: !1,
            ordering: !0,
            alphabet: {
                column: 3,
                caseSensitive: true,
                numbers: true,
            },
            order: [],
            rowGroup: "",
            info: !0,
            procesing: !0,
            select: !0,
            searching: !0,
            stateSave: !0,
            paging: !0,
            pagingType: ["full_numbers"],
            lengthMenu: [
                [15, 25, 50, 100, -1],
                [15, 25, 50, 100, "All"]
            ],
            buttons: [
                { extend: "copyHtml5" },
                {
                    extend: "excelHtml5",
                    exportOptions: {
                        columns: ":visible",
                        format: {
                            body: function (data, row, column, node) {
                                return data = $("<p>" + data + "<\/p>").text(), $.isNumeric(data.replace(",", ".")) ? data.replace(",", ".") : data
                            }
                        }
                    }
                },
                { extend: "csvHtml5" },
                {
                    extend: "pdfHtml5",
                    pageSize: "A3",
                    orientation: "landscape"
                },
                { extend: "pageLength" }
            ],
            responsive: { details: { type: "inline" } }
        })
    });
</script>
```

Change is MIT - feel free to use as you want.

### CODE REVIEW

**Comment by DataTables** [comment]
Thanks for the PR - this is an excellent enhancement to the software! However, I'm afraid I don't think it is quite right. Consider the case where you have multiple tables on the page - each time the software is initialised this runs:

```
$.fn.dataTable.AlphabetSearch.column = options.column;
```

that assigns a static parameter which is shared between all instances. So if you were to have one table with column 0 and another with column 1, then attempt to redraw the first, it would actually be referencing column index 1!

What needs to happen is to refer to the `options` object through out each instance, so the values can be unique per DataTable.

**Comment by PrzemyslawKlys** [comment]
Hi,

I believe I have tested this scenario and it worked fine:
- https://jsfiddle.net/MadBoy/3j7mzndo/

It covers 4 tables with the same data, just different alphabet settings and it seems to work correctly. 

I'm using the "current PR" code for the alphabet and it seems fine. 

```html
        <link href="https://cdn.jsdelivr.net/gh/PrzemyslawKlys/Plugins@master/features/alphabetSearch/dataTables.alphabetSearch.css" as="style" type="text/css" rel="stylesheet preload prefetch" />
        <script src="https://cdn.jsdelivr.net/gh/PrzemyslawKlys/Plugins@master/features/alphabetSearch/dataTables.alphabetSearch.js"></script>
```

I believe the situation you are referring to won't happen as the code ensures that if options are not set it's set to column 0, caseSensitive false, and numbers to false (as in use defaults - for current settings). If the user overwrites it uses current settings.

```js
	var options = $.extend({
		column: 0,
		caseSensitive: false,
		numbers: false
	}, table.init().alphabet);

	$.fn.dataTable.AlphabetSearch.column = options.column;
	$.fn.dataTable.AlphabetSearch.caseSensitive = options.caseSensitive;
	$.fn.dataTable.AlphabetSearch.addNumbers = options.numbers;
```

I could try to fixing it without using it that way - just I am a bit noobish on JS so I am not sure on scoping this correctly and refreshing options.column

File: features/alphabetSearch/dataTables.alphabetSearch.js [hunk 1]
```diff
@@ -49,36 +49,44 @@ $.fn.dataTable.ext.search.push( function ( context, searchData ) {
...
-    if ($.fn.dataTable.AlphabetSearch.caseSensitive) {
-        if ( searchData[$.fn.dataTable.AlphabetSearch.column].charAt(0) === context.alphabetSearch ) {
-            return true;
-        }
-    } else {
-        if ( searchData[$.fn.dataTable.AlphabetSearch.column].charAt(0).toUpperCase() === context.alphabetSearch ) {
-            return true;
-        }
-    }
+	if (context.oInit.alphabet !== undefined) {
```
Comment (PrzemyslawKlys) [inline features/alphabetSearch/dataTables.alphabetSearch.js:52]: I wasn't sure how to get variables set in alphabet visible and this seemed like the only way to do it. If there's better way please let me know how.

File: features/alphabetSearch/dataTables.alphabetSearch.js [hunk 2]
```diff
@@ -184,9 +197,6 @@ $.fn.dataTable.AlphabetSearch = function ( context ) {
 	};
 };
 
-$.fn.DataTable.AlphabetSearch = $.fn.dataTable.AlphabetSearch;
```
Comment (PrzemyslawKlys) [inline features/alphabetSearch/dataTables.alphabetSearch.js:187]: I am not sure why this was there at all. Removed it as I believe it's some kind of mistake.

**Comment by PrzemyslawKlys** [comment]
Ok, can you review it now? I believe I've addressed the issue.  Please let me know if you think this will work?

File: features/alphabetSearch/dataTables.alphabetSearch.js (hunk 2 above)
Comment (DataTables) [reply features/alphabetSearch/dataTables.alphabetSearch.js:187]: Actually it is needed - `$.fn.dataTable` and `$.fn.DataTable` are actually different objects - but for ease we tend to make what is available on one, available on the other.

**Comment by DataTables** [comment]
I think that is looking really good. Just that one tweak above and we'll call it done :-)

File: features/alphabetSearch/dataTables.alphabetSearch.js (hunk 2 above)
Comment (PrzemyslawKlys) [reply features/alphabetSearch/dataTables.alphabetSearch.js:187]: Fixed

**Comment by PrzemyslawKlys** [comment]
Should be good now.

**Comment by DataTables** [comment]
Thanks for taking the time to do this - it is a really nice addition to the plugin!

**Comment by PrzemyslawKlys** [comment]
Cool. What's the release cycle now? When is .min.js gonna be generated, version bumped and so on?

**Comment by DataTables** [comment]
If not by the end of February, then in March. We don't have a set release date, but I am keep to get the Bootstrap 5 support shipped.

**Pull request merged**